### PR TITLE
Unpaged partitioning

### PR DESCRIPTION
**Pull request by jenswi-linaro**
This removes about 7 kB of unpaged code and data from QEMU virt.

### CODE REVIEW

File: core/arch/arm/kernel/secure_partition.c [hunk 1]
```diff
@@ -349,7 +349,7 @@ static bool sp_handle_svc(struct thread_svc_regs *regs)
 	return false;
 }
 
-static const struct ts_ops sp_ops __rodata_unpaged = {
+static const struct ts_ops sp_ops __rodata_unpaged("sp_ops") = {
```
Comment (jforissier) [inline core/arch/arm/kernel/secure_partition.c]: Could you avoid repeating the variable name? (I see a potential difficulty with checkpatch but we may ignore it)

File: core/arch/arm/kernel/link_dummies_paged.c [hunk 2]
```diff
@@ -38,3 +40,18 @@ __wq_rpc(uint32_t func __unused, int id __unused,
...
+const struct mobj_ops mobj_mm_ops __rodata_dummy;
+const struct mobj_ops mobj_shm_ops __rodata_dummy;
+const struct mobj_ops mobj_seccpy_shm_ops __rodata_dummy;
+const struct mobj_ops mobj_with_fobj_ops __rodata_dummy;
+const struct fobj_ops ops_rwp_paged_iv __rodata_dummy;
+const struct fobj_ops ops_rwp_unpaged_iv __rodata_dummy;
+const struct fobj_ops ops_ro_paged __rodata_dummy;
+const struct fobj_ops ops_ro_reloc_paged __rodata_dummy;
+const struct fobj_ops ops_locked_paged __rodata_dummy;
+const struct fobj_ops ops_sec_mem __rodata_dummy;
```
Comment (jforissier) [inline core/arch/arm/kernel/link_dummies_paged.c:52]: Should be a separate commit

File: core/arch/arm/mm/mobj_ffa.c [hunk 3]
```diff
@@ -39,7 +39,7 @@ static struct mobj_ffa_head shm_inactive_head =
 
 static unsigned int shm_lock = SPINLOCK_UNLOCK;
 
-static const struct mobj_ops mobj_ffa_ops __rodata_unpaged("mobj_ffa_ops");
+const struct mobj_ops mobj_ffa_ops;
```
Comment (jforissier) [inline core/arch/arm/mm/mobj_ffa.c:42]: I don't understand this change, it is not what the commit description says

File: core/mm/fobj.c [hunk 4]
```diff
@@ -55,8 +55,8 @@ struct fobj_rwp_paged_iv {
 	struct fobj fobj;
 };
 
-static const struct fobj_ops ops_rwp_paged_iv;
-static const struct fobj_ops ops_rwp_unpaged_iv;
+const struct fobj_ops ops_rwp_paged_iv;
+const struct fobj_ops ops_rwp_unpaged_iv;
```
Comment (jforissier) [inline core/mm/fobj.c:60]: I don't understand this change, it is not what the commit description says

File: core/mm/fobj.c [hunk 5]
```diff
@@ -408,7 +416,7 @@ struct fobj_rop {
 	struct fobj fobj;
 };
 
-static const struct fobj_ops ops_ro_paged;
+const struct fobj_ops ops_ro_paged;
```
Comment (jforissier) [inline core/mm/fobj.c:419]: I don't understand this change, it is not what the commit description says

File: core/mm/fobj.c [hunk 6]
```diff
@@ -656,15 +668,19 @@ static TEE_Result rrp_load_page(struct fobj *fobj, unsigned int page_idx,
...
+const struct fobj_ops ops_ro_reloc_paged
+__weak __rodata_unpaged("ops_ro_reloc_paged") = {
 	.free = rrp_free,
 	.load_page = rrp_load_page,
 	.save_page = rop_save_page, /* Direct reuse */
 };
 #endif /*CFG_CORE_ASLR*/
 
-static const struct fobj_ops ops_locked_paged;
+const struct fobj_ops ops_locked_paged;
```
Comment (jforissier) [inline core/mm/fobj.c:683]: I don't understand this change, it is not what the commit description says

File: core/mm/mobj.c [hunk 7]
```diff
@@ -522,7 +542,7 @@ struct mobj_with_fobj {
 	struct mobj mobj;
 };
 
-static const struct mobj_ops mobj_with_fobj_ops;
+const struct mobj_ops mobj_with_fobj_ops;
```
Comment (jforissier) [inline core/mm/mobj.c:545]: I don't understand this change, it is not what the commit description says

**Review by jforissier — COMMENTED** [review]
Have you tested with Clang too? (I haven't -- just asking in case compiler/linker would behave slightly differently)

File: core/arch/arm/kernel/secure_partition.c (hunk 1 above)
Comment (jenswi-linaro) [reply core/arch/arm/kernel/secure_partition.c]: Yes, but then we'd have a macro that changes the syntax and that's in this case even worse in my opinion.

File: core/arch/arm/kernel/link_dummies_paged.c (hunk 2 above)
Comment (jenswi-linaro) [reply core/arch/arm/kernel/link_dummies_paged.c:52]: I'm sorry I don't get it. What should be a separate commit?

File: core/arch/arm/mm/mobj_ffa.c (hunk 3 above)
Comment (jenswi-linaro) [reply core/arch/arm/mm/mobj_ffa.c:42]: static variables cannot use `__weak` so we must make this non-static, that's mentioned in the commit. Perhaps I should have made this line `extern struct mobj_ops mobj_ffa_ops` instead, but then checkpatch would have complained.
Yes, I'm dropping `__rodata_unpaged("mobj_ffa_ops")` but it's still kept below so that part hasn't really changed. I didn't mention this in the commit since I didn't think it was significant.

**Comment by jenswi-linaro** [comment]
> Have you tested with Clang too? (I haven't -- just asking in case compiler/linker would behave slightly differently)

No

File: core/arch/arm/mm/tee_pager.c [hunk 8]
```diff
@@ -1027,8 +1027,6 @@ static void rem_region(struct vm_paged_region_head *regions,
 	}
 
 	pager_unlock(exceptions);
-
-	free_region(reg);
 }
 DECLARE_KEEP_PAGER(rem_region);
```
Comment (etienne-lms) [inline core/arch/arm/mm/tee_pager.c:1031]: I wonder why rem_region() needs to be unpaged since its called only from tee_pager_rem_um_region() which has no pager constraints.

File: core/arch/arm/kernel/link_dummies_paged.c [hunk 9]
```diff
@@ -32,9 +35,21 @@ __thread_std_smc_entry(uint32_t a0 __unused, uint32_t a1 __unused,
...
 	return 0;
 }
-void __section(".text.dummy.__wq_rpc")
-__wq_rpc(uint32_t func __unused, int id __unused,
-	 const void *sync_obj __unused, const char *fname __unused,
-	 int lineno  __unused)
-{
-}
+
+const struct mobj_ops mobj_reg_shm_ops __rodata_dummy;
```
Comment (etienne-lms) [inline core/arch/arm/kernel/link_dummies_paged.c:39]: minor: what about using alphabetical ordering for this definition series.

**Review by etienne-lms — COMMENTED** [review]
Commit: "core: separate stmm_sp_ops using a __weak attribute instead" required updated in _stmm_sp.h_:
```
--- a/core/arch/arm/include/kernel/stmm_sp.h
+++ b/core/arch/arm/include/kernel/stmm_sp.h
@@ -140,12 +140,9 @@ struct stmm_ctx {
        bool is_initializing;
 };
 
-extern const struct ts_ops *stmm_sp_ops_ptr;
-
 static inline bool is_stmm_ctx(struct ts_ctx *ctx __maybe_unused)
 {
-       return IS_ENABLED(CFG_WITH_STMM_SP) && stmm_sp_ops_ptr &&
-              ctx && ctx->ops == stmm_sp_ops_ptr;
+       return IS_ENABLED(CFG_WITH_STMM_SP) && ctx && ctx->ops == stmm_sp_ops;
 }
 
 static inline struct stmm_ctx *to_stmm_ctx(struct ts_ctx *ctx)
```
.

For the few last commit, it looks strange to see inline comment like:
```
 * Note: this variable is weak just to make it possible to exclude it
 * from the unpaged area.
```
specifically on resource tagged to be included in an unpaged section.
Maybe change to something like below (reusing your wording in commit logs):
```
 * Note: this variable is weak just to ease breaking its dependency chain
 * when added to the unpaged area.
```

File: core/arch/arm/mm/tee_pager.c (hunk 8 above)
Comment (jenswi-linaro) [reply core/arch/arm/mm/tee_pager.c:1031]: Because it holds the pager lock.

**Comment by etienne-lms** [comment]
Note: I edited the proposed patch above, it was bugged.

File: core/arch/arm/kernel/link_dummies_paged.c (hunk 9 above)
Comment (jenswi-linaro) [reply core/arch/arm/kernel/link_dummies_paged.c:39]: Or perhaps sorting it based on line length? (Just kidding) :-)
That's a bit much in my opinion, the order of these has no impact.

Comment (etienne-lms) [reply core/arch/arm/kernel/link_dummies_paged.c:39]: ok :)

**Comment by jenswi-linaro** [comment]
Updated the comments as suggest

File: core/arch/arm/kernel/secure_partition.c (hunk 1 above)
Comment (jforissier) [reply core/arch/arm/kernel/secure_partition.c]: Fair enough, let's keep this.

File: core/arch/arm/kernel/link_dummies_paged.c (hunk 2 above)
Comment (jforissier) [reply core/arch/arm/kernel/link_dummies_paged.c:52]: Sorry I reviewed the patches one by one but I submitted all comments at once so the context is unclear.
You are introducing this change in commit "core: make __rodata_unpaged() symbols __weak". This does not fit the description.

File: core/arch/arm/mm/mobj_ffa.c (hunk 3 above)
Comment (jforissier) [reply core/arch/arm/mm/mobj_ffa.c:42]: Oh! It is a forward declaration. How did I miss that? :confused:
Sorry that's fine then.

File: core/arch/arm/kernel/link_dummies_paged.c (hunk 2 above)
Comment (jenswi-linaro) [reply core/arch/arm/kernel/link_dummies_paged.c:52]: The part I'm not getting is how it doesn't fit the description in "core: make __rodata_unpaged() symbols __weak".

Comment (jforissier) [reply core/arch/arm/kernel/link_dummies_paged.c:52]: Ah yes, "in order to be overridden". Sorry, really not my day :cry: I think I'll stop reviewing stuff for now :roll_eyes: :laughing:

Comment (jenswi-linaro) [reply core/arch/arm/kernel/link_dummies_paged.c:52]: No worries, we all have days like that. :-)

File: core/arch/arm/mm/tee_pager.c (hunk 8 above)
Comment (etienne-lms) [reply core/arch/arm/mm/tee_pager.c:1031]: indeed, right!

**Comment by jforissier** [comment]
> > Have you tested with Clang too? (I haven't -- just asking in case compiler/linker would behave slightly differently)
> 
> No

Build tested OK with Clang and pager enabled, but there is an unrelated crash on boot (https://github.com/OP-TEE/optee_os/issues/4621).

**Comment by jforissier** [comment]
`Tested-by: Jerome Forissier <jerome@forissier.org> (QEMU, pager, Clang 12)`

**Comment by jenswi-linaro** [comment]
Squashed and rebased on master. Applied the T-B tag.

File: core/arch/arm/kernel/link_dummies_paged.c (hunk 9 above)
Comment (jforissier) [reply core/arch/arm/kernel/link_dummies_paged.c:39]: Why does it matter to put all these symbols in `.rodata.dummy`? In other words, what happens if `__rodata_dummy` is not added?

**Review by jforissier — COMMENTED** [review]
- For commits "core: pager: don't call free_region() from paged function",
     "core: use separate sections for each __rodata_unpaged variable"
    `Reviewed-by: Jerome Forissier <jerome@forissier.org>`
- For commits "...using a __weak attribute instead":
With my previous comment about the use __rodata_unpaged answered or addressed:
    `Reviewed-by: Jerome Forissier <jerome@forissier.org>`
- For commit "core: make __wq_rpc() static again"
    `Acked-by: Jerome Forissier <jerome@forissier.org>`

File: core/arch/arm/kernel/link_dummies_paged.c (hunk 9 above)
Comment (jenswi-linaro) [reply core/arch/arm/kernel/link_dummies_paged.c:39]: Then they'd end up with the original section name and get pulled into the unpaged section for that reason. Now for these structs it's wouldn't be so bad since we're anyway forcing them into the unpaged section, but I still think it's important that it ends up in the unpaged section for the right reason.

Comment (jforissier) [reply core/arch/arm/kernel/link_dummies_paged.c:39]: Oh, to be honest I had somewhat forgotten (again) about the link process. Got it, thanks.

**Comment by jenswi-linaro** [comment]
Tags applied.

File: core/arch/arm/kernel/link_dummies_paged.c (hunk 9 above)
Comment (jenswi-linaro) [reply core/arch/arm/kernel/link_dummies_paged.c:39]: > Oh, to be honest I had somewhat forgotten (again) about the link process. Got it, thanks.

Understandable, it is complicated. The good news is that it usually work well enough to be temporarily forgotten. :-)

**Comment by jenswi-linaro** [comment]
Tag applied

**Comment by jforissier** [comment]
> Note that the commit "core: make __rodata_unpaged() symbols __weak" still need some review.

For that commit:
`Reviewed-by: Jerome Forissier <jerome@forissier.org>`

**Comment by jenswi-linaro** [comment]
Thanks @jforissier. Tag applied.

**Comment by etienne-lms** [comment]
Fyi, i'm fine with the content. looks all good to me.
But since i'm stucked in https://github.com/OP-TEE/optee_os/pull/4628, i'm not that sure i can tell it's enough tested.
`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>`

**Comment by jenswi-linaro** [comment]
Let's take #4628 first and we can test again.

**Comment by jforissier** [comment]
@etienne-lms now that #4628 is merged, could you please test this rebased onto master? Thanks!

**Comment by jenswi-linaro** [comment]
Rebased on master.

**Comment by jforissier** [comment]
@jenswi-linaro could you please apply @etienne-lms' Ack? With the 3.14.0 release approaching, I'd rather take this sooner than later.

**Comment by jenswi-linaro** [comment]
@etienne-lms' ack applied.
